pkg/kv/kvserver/store_raft.go

```diff
@@ -1138,3 +1337,334 @@ func (s *Store) updateCapacityGauges(ctx context.Context) error {
 
 	return nil
 }
+
+// raftSchedulerForDequeuePacer abstracts raftScheduler for the
+// storeRaftDequeuePacer.
+type raftSchedulerForDequeuePacer interface {
+	EnqueueRaftRequest(id roachpb.RangeID)
+}
+
+// engineForDequeuePacer abstracts storage.Engine for the
+// storeRaftDequeuePacer.
+type engineForDequeuePacer interface {
+	// TryWaitForMemTableStallHeadroom return true if the number of engine
+	// memtables is small enough to have some headroom to avoid write stalls, in
+	// which case it also returns the allowedBurst in bytes (which must be
+	// greater than zero). Otherwise, it returns false and allowedBurst is zero.
+	//
+	// If doWait is true, it waits until there is headroom before returning, in
+	// which case the return value is guaranteed to be true.
+	//
+	// This is not optimized to be very efficient, so it is preferable that it
+	// is only called by a few callers, who cache the result for a short
+	// duration (say 10ms).
+	TryWaitForMemTableStallHeadroom(doWait bool) (ok bool, allowedBurst int64)
```

Comment:
Since you are abstracting here, does this need to have such a specific name mentioning memtables etc? Could this be something like `TryWaitFor[Stall]Headroom`?

Or the plan is to de-`interface` this and use `storage.Engine` directly when it implements this method?